### PR TITLE
improve latex caption injection

### LINKED ISSUE #4748: Long-table cannot have a cross-referenced label or tbl-cap when using kableExtra col_spec
### Bug description

Hi, and thanks for a highly usable tool; quarto.

I have a couple of related issues with a table made by `kableExtra::kbl()` and `kableExtra::kable_styling()` which spans multiple pages in the resulting pdf.

### First issue
 
When I render the pdf in quarto with a label `#| label: tbl-longTable` the table are cut on the first page. 

If I instead remove the label the whole table is rendered spanning multiple pages but of course the cross reference in the text does not work.

### Second issue

If I use a `#| tbl-cap: ` for the same table I get following error while rendering the report.

``````
Package array Error:  Illegal pream-token (\caption): `c' used.

``````

However, if I do not have the label the table is generated but without caption.

Also with the lable the R-code is not showing in the pdf. 

I have traced the issues to when you add the `column_spec()`, without them the table renders fine with both *label:* and *tbl-cap:*.
Everything does work with the `row_spec()` however.

Any Ideas of how to proceed?

--------------

**R-studio**

```
RStudio 2022.12.0+353 "Elsbeth Geranium"
Release (7d165dcfc1b6d300eb247738db2c7076234f6ef0, 2022-12-03) for Ubuntu Jammy
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)
rstudio/2022.12.0+353 Chrome/102.0.5005.167 Electron/19.1.3 Safari/537.36
```

**Linux**

`Linux Mint 21.1, kernel 5.15.0-67-generic`

**Quarto**
`quarto 1.2.335`

------------------------

### A .qmd with example code

Below is a reprex with the YAML and R-code which reproduce my error on my system.

````
---
title: "reprex_kable"
format: pdf
documentclass: scrreprt
execute: 
  warning: false
  error: false
  message: false
---

The table @tbl-cars should be a table spaning more than one page, with a caption.


```{r}
#| label: tbl-cars
#| echo: false
#| cache: true
#| tbl-cap: "A table with many models repeating itself"

library(tidyverse)
library(kableExtra)

mtcars %>% 
  bind_rows(mtcars, mtcars) %>% 
  kableExtra::kbl(booktabs = T, longtable = T) %>% 
  kableExtra::kable_styling(latex_options = c("hold_position", "striped"), full_width = F, repeat_header_continued = T, repeat_header_text = "Continuation table:") %>% 
  column_spec(column = 1:2, width = "4cm") %>%
  column_spec(column = 1, bold = T) %>%
  row_spec(row = 0, bold = T, background = "#3e77c1", font_size = 12, color = "white", hline_after = F) %>%
  row_spec(1:107, hline_after = T, font_size = 10)

```
````

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [X] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

### PR DESCRIPTION
**Pull request by cscheid**
Closes #4748.

This isn't a true solution because it will only work if the originating latex includes all table specifiers in a single line. Unfortunately, we can't do better than that without a context-sensitive parser, because in the _actual good solution_, we need to match balanced curly brackets.

> Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems.

:) Oh well.

### CODE REVIEW

**Comment by cderv** [comment]
We should check also the initial issue from #1104 with **gt** to see if that fixes it too and do not introduce a regression. Did you do that already ? 

There wasn't smoke all at the time so not as easy as now to add test.

**Comment by cscheid** [comment]
@cderv good point. I checked locally but now I've added it to the smoke tests.

**Comment by cderv** [comment]
I was fearing this smoke test failure on Ubuntu because of renv update as all RSPM were replaced by CRAN source. On Ubuntu GHA, installing from RSPM is better to have linux binary. I need to manage to for the config somehow to RSPM as it seems locally your install uses CRAN only and updated the renv lockfile accordingly, including package you did not change. I was not expecting that... 🤔

**Comment by cderv** [comment]
Also I wonder why I did not go with `\n` in the regex ... 🤔 I wonder if this is because of Windows newline being different that Unix newline. Smoke test will tell us.

**Comment by cscheid** [comment]
@cderv I picked `\n` because Windows linebreaks are `\r\n` so that should still work (we'll match to the last `}` before `\n`). It won't work on MacOS 9 which uses `\r` only 🤣

File: .github/workflows/test-smokes.yml [hunk 1]
```diff
@@ -87,6 +87,9 @@ jobs:
           if (!requireNamespace('renv', quietly = TRUE)) install.packages('renv')
           renv::restore()
         shell: Rscript {0}
+        env:
+          RENV_CONFIG_REPOS_OVERRIDE: https://packagemanager.rstudio.com/cran/latest
+          RENV_CONFIG_CACHE_ENABLED: FALSE
```
Comment (cderv) [inline .github/workflows/test-smokes.yml:92]:
```suggestion
        env:
          RENV_CONFIG_REPOS_OVERRIDE: https://packagemanager.rstudio.com/cran/latest
```
We only need this I think

Comment (cscheid) [reply .github/workflows/test-smokes.yml:92]: Yup. I'll wait for the slow one to pass first, though.

**Comment by cderv** [comment]
> I picked `\n` because Windows linebreaks are `\r\n` so that should still work (we'll match to the last `}` before `\n`). It won't work on MacOS 9 which uses `\r` only 🤣

Ah damn... does Lua support the usual  `(\r\n?|\n)`  that we use in Typescript ? Maybe not... 🤔

**Comment by cscheid** [comment]
No, Lua patterns do not support `|` (this is a huge pain point for us in general).

**Comment by cderv** [comment]
:( thanks for confirming. That is what I thought but was unsure... very painful. 

Maybe there is a lightweight [Pattern matching libraries](http://lua-users.org/wiki/LibrariesAndBindings) that we'll need to use in the future as we do with other lib...  but usually they are rather heavy.

It seems like **renv** installation went ok with the env var. cool !

**Comment by cscheid** [comment]
@cderv Yes, I'd like to use a PCRE library like https://luarocks.org/modules/rrt/lrexlib-pcre, but they depend on dynamic linking, and Pandoc's version of Lua doesn't support that :(

**Comment by cderv** [comment]
Oh some new stuff I learn. Did not know that limitation. Thanks.  We'll be limited for now then 😓

**Pull request merged**